### PR TITLE
Build & test code on Travis CI

### PR DESCRIPTION
**Pull request by Leonidas-from-XIV**
I've been trying to get Travis to build the code (mostly successfully) and run the code (unsucessfully). I don't quite know why it is failing to connect. I traced one of the issues down to disabling IPv6, but maybe you have an idea on what the connection issue might be.

There is some other packaging/dependency problem that I also need to look into but the problem with connectivity is what makes this really annoying.

Edit: [Here](https://travis-ci.org/Leonidas-from-XIV/ocaml-kafka/builds) is a list of builds where I tried to coerce Travis to do the right thing.

### CODE REVIEW

**Comment by didier-wenzek** [comment]
I don't think this is related to ipv6. I tried both settings following, your script and both are working.

The issue is related to the name `localhost`. Removing the line `127.0.0.1 localhost` from `/etc/hosts` breaks the tests with the very same message you have with Travis. This is the case, even with the changes you pushed with an explicit call to `127.0.0.1`.

```
dune runtest
       tests alias test/runtest (exit 2)
(cd _build/default/test && ./tests.exe)
%3|1579288631.971|FAIL|rdkafka#producer-3| [thrd:localhost:9092/bootstrap]: localhost:9092/bootstrap: Connect to ipv6#[::1]:9092 failed: Connection refused (after 5ms in state CONNECT)
%3|1579288631.971|ERROR|rdkafka#producer-3| [thrd:localhost:9092/bootstrap]: localhost:9092/bootstrap: Connect to ipv6#[::1]:9092 failed: Connection refused (after 5ms in state CONNECT)
%3|1579288631.971|ERROR|rdkafka#producer-3| [thrd:localhost:9092/bootstrap]: 1/1 brokers are down
Fatal error: exception File "test/tests.ml", line 181, characters 3-9: Assertion failed
Makefile:11: recipe for target 'test' failed
make: *** [test] Error 1
```

**Comment by Leonidas-from-XIV** [comment]
Thanks for your help with reproducing. If I look at the `/etc/hosts` on Travis it looks pretty much *normal*:

```
## Managed by Chef on packer-5e046b3e-10f1-6b54-7282-1ce424fb447b.c.eco-emissary-99515.internal :heart_eyes_cat:
## cookbook:: travis_build_environment
##     file:: templates/default/etc/cloud/templates/hosts.tmpl.erb
127.0.1.1 localhost localhost ip4-loopback 
127.0.0.1 localhost nettuno travis vagrant travis-job-812ffc5e-7344-4251-8d26-cc83f4e212c9 
```

The somewhat strange `127.0.1.1` IP is some Debian quirk to fix other quirks somewhere else.

**Comment by didier-wenzek** [comment]
Thanks for persevering!

Did the change to remove the weird entry in `/etc/hosts` fixed anything? Looking at the Travis logs it seems that not. However, I didn't find logs for the `before_install` scripts.

**Comment by didier-wenzek** [comment]
Sorry, my previous comment has been written a bit too quickly ... I found the logs I was looking for.
However, I noticed nothing wrong.

`advertised.listeners=null` but this is should not be a blocker.

**Comment by Leonidas-from-XIV** [comment]
I have finally figured out where the issue is: opam 2.0.4+ does sandboxing and disallows code that is running under it (like `dune runtest`) to connect to Kafka.

I have moved out the tests from the `runtest` to an `integration` target, that should only be called on CI and on developers machines, but not when calling opam to run tests. Which is a good thing, since otherwise the tests will fail when submitted in `opam-repository`, otherwise they would depend on `opam-repository` disabling sandboxing and installing Kafka to test.

I think this is a passable compromise (the same I do on `orewa`, whose integration tests similarly depend on Redis), I'll just work out the kinks between `kafka_lwt` and `kafka` packages now.

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+matrix:
+  include:
+  - os: linux
+    env: OCAML_VERSION=4.02 PACKAGE="kafka"     PINS="kafka:."
+  - os: linux
+    env: OCAML_VERSION=4.02 PACKAGE="kafka_lwt"
+  - os: linux
+    env: OCAML_VERSION=4.09 PACKAGE="kafka"     PINS="kafka:."
+  - os: linux
+    env: OCAML_VERSION=4.00 PACKAGE="kafka_lwt" PINS="kafka:. kafka_lwt:."
```
Comment (didier-wenzek) [inline .travis.yml]: This should be 4.09. The corresponding build fails to compile `dune`. These are likely related.

**Comment by didier-wenzek** [comment]
Nice to see the Travis tests running on Mac! Thanks!

`dune runtest` hangs on my laptop now. I will take the time to understand why later.

**Comment by didier-wenzek** [comment]
Making a distinction between unit and integration tests is great. However, there are small issues here:
* There are no *true* unit tests so far.
* `issue3.ml` is *not* an automated test, but an executable I wrote to reproduce the [issue #3](https://github.com/didier-wenzek/ocaml-kafka/issues/3). This is an infinite consumer loop. 
* So `dune runtest` cannot trigger `issue3.exe`  as done in `test_lwt/dune `.

What I propose to move forward:
* As you have already done, run the former tests under `dune integration`.
* Write a minimal unit test suite to run under `dune test`. This can even be empty to start.

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    - sleep 3
+    - kafka/bin/kafka-server-start.sh kafka/config/server.properties &
+    - sleep 5
+    - kafka/bin/kafka-topics.sh --create --bootstrap-server localhost:9092 --replication-factor 1 --partitions 2 --topic test
+matrix:
+  include:
+  - os: linux
+    env: OCAML_VERSION=4.02 PACKAGE="kafka"     PINS="kafka:."
+  - os: linux
+    env: OCAML_VERSION=4.02 PACKAGE="kafka_lwt" PINS="kafka:. kafka_lwt."
```
Comment (didier-wenzek) [inline .travis.yml]: There is a missing semicolon `:` in `kafka_lwt.`. Should be `kafka_lwt:.`.

**Comment by didier-wenzek** [comment]
I saw your comment on OCaml discuss. Unfortunately, I have no clue why the behavior differs for the two packages. What I find weird is that the version constraint translates to ` kafka_lwt → kafka < 0.2`. 

Removing this version constraint fixes the `opam pin add kafka_lwt --kind=path .` command on my host.

```
git diff kafka_lwt.opam

-  "kafka" {= version}
+  "kafka"
```

**Comment by Leonidas-from-XIV** [comment]
Okay, after getting help from @avsm on the OCaml discuss and sorting out the remaining issues (it picked up a too old version of Alcotest, the Lwt tests being built in the non-Lwt context), I think I [got it to work](https://travis-ci.org/Leonidas-from-XIV/ocaml-kafka/builds/642372042).

The issue is that the integration tests times out, which is also the case on my machine. But that was always the case, even before we had the PR, so this is now a good time to figure out why. @didier-wenzek, could you take a look what is going on?

**Comment by didier-wenzek** [comment]
@Leonidas-from-XIV The integration tests are running on my laptop (Ubuntu 18.04.3 LTS) are running without any issues (before and after your changes), _but_ I run a newer version of `librdkafka`.

* The tests are ok, with the version [1.0.0](https://github.com/edenhill/librdkafka/releases/tag/v1.0.0)
* The tests failed with a timeout for the version [0.11.3 distributed by ubuntu](https://packages.ubuntu.com/source/bionic/librdkafka)

I have to:
* find version of `librdkaka` broke the tests and why
* check that the tests are ok for the very last version `1.3.0`.

**Comment by Leonidas-from-XIV** [comment]
I am not sure this is a problem with the version. I am running the integration test on macOS, with librdkafka-1.3.0 and they always get stuck. The same version on Travis seems to run fine on macOS with the same librdkafka installed via brew. I'll investigate which step it is that the integration test is hanging on.

The 0.11.3 version shipped in Ubuntu might in fact be ancient, if this is the problem it is good to know about it, then we can install newer versions of it and will be aware of the minimum version requirement.

**Comment by didier-wenzek** [comment]
I think this related to the version. I checked different versions of `librdkafka`:

* The integration tests are ok for all the versions from [1.0.0.RC2](https://github.com/edenhill/librdkafka/releases/tag/v1.0.0-RC2) up to [1.3.0](https://github.com/edenhill/librdkafka/releases/tag/v1.3.0) 
* There is no 1.0.0.RC1 and the previous version was [0.11.6](https://github.com/edenhill/librdkafka/releases/tag/v0.11.6). The tests are failing on an `Assert_failure("test/integration/integration.ml", 181, 3)`
* The version installed with ubuntu [0.11.3](https://github.com/edenhill/librdkafka/releases/tag/v0.11.3) leads to a time out in the same place related to producers' delivery callbacks.

We might deprecate versions before 1.0.0. However, I will first take the time to understand what is going on. It's in particular bizarre that the tests get stuck on your host. Are you sure to exercice the version 1.3.0? In my case, the version installed by ubuntu is found by `ldd` before the one I installed in `/usr/lib`. 

Unless you plan to update your scripts to use the version 1.0.0 of librdkafka, I will merge this pull request, the issue being unrelated.
 
For

**Comment by Leonidas-from-XIV** [comment]
> Unless you plan to update your scripts to use the version 1.0.0 of librdkafka, I will merge this pull request, the issue being unrelated.

Yes, please do for now. I looked around for 1.0+ versions of librdkafka packaged in PPAs that I could use easily, but I couldn't find any so I would need to create a PPA myself. That will take a while so I think this branch can be merged meanwhile to have some kind of CI integration in master and then we can iterate on it in a separate branch.

**Comment by Leonidas-from-XIV** [comment]
I think you still need to go to Travis https://travis-ci.org/didier-wenzek/ocaml-kafka and click around to enable building it somehow.

**Pull request merged**